File: app/src/main/java/com/victoria/bleled/util/arch/base/BaseViewModel.java
package com.victoria.bleled.util.arch.base;


import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.victoria.bleled.util.arch.network.NetworkResult;


public class BaseViewModel extends ViewModel {
    // one direction
    protected MutableLiveData<Boolean> _dataLoading = new MutableLiveData<Boolean>();
    public LiveData<Boolean> dataLoading = _dataLoading;

    protected MutableLiveData<String> _toastMessage = new MutableLiveData<String>();
    public LiveData<String> toastMessage = _toastMessage;

    protected MutableLiveData<NetworkResult> _networkErrorLiveData = new MutableLiveData<>();
    public LiveData<NetworkResult> networkErrorLiveData = _networkErrorLiveData;

    public BaseViewModel() {
        super();
    }

}

File: app/src/main/java/com/victoria/bleled/base/BaseViewModel.kt
package com.victoria.bleled.base

import androidx.annotation.StringRes
import androidx.lifecycle.ViewModel
import com.victoria.bleled.base.internal.AppException
import com.victoria.bleled.base.internal.SingleLiveEvent
import io.reactivex.disposables.CompositeDisposable
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import timber.log.Timber


open class BaseViewModel : ViewModel() {

    var _error = SingleLiveEvent<Exception>()
    var _alert = SingleLiveEvent<Any>()

    fun error(exception: AppException) {
        _error.value = exception
    }

    fun alert(@StringRes message: Int) {
        _alert.value = message
    }

    fun alert(message: String?) {
        if (message == null) {
            Timber.w("Alert Message Empty!!")
            // Nothings..
        } else {
            _alert.value = message!!
        }
    }

    override fun onCleared() {
        //disposebag.dispose()
        super.onCleared()
    }

    inline fun CoroutineScope.launchInSafe(
        crossinline error: ((Throwable) -> Unit) = {},
        crossinline handle: ((AppException.Handle) -> Unit) = {},
        crossinline block: suspend CoroutineScope.() -> Unit,
    ): Job {
        return this.launch(CoroutineExceptionHandler { _, throwable ->
            error(throwable)
            throwable.printStackTrace()
            when (throwable) {
                is AppException.ServerHttp -> {
                    _error.value = throwable
                }
                is AppException.Network -> {
                    _error.value = throwable
                }
                is AppException.Handle -> {
                    handle(throwable)
                }
                else -> {
                    _error.value = AppException.Unknown(throwable)
                }
            }
        }) {
            block()
        }
    }
}